### PR TITLE
[nomerge] optimise the addition of immutable HashMap

### PR DESCRIPTION
**Pull request by mkeskells**
Note - for 2.12 only
We can duplicate the functionality to 2.13, but I think it is with a seperate PR as 2.13 collections are differently structured

Prefer to use `merge` when we can
optimised the builder to allow `++` directly

small collection of other optimisations

need some additional tests for the edge cases introduced

### CODE REVIEW

**Comment by retronym** [comment]
Minimization of the bootstrap failure:

```
scala> val breaker = collection.breakOut[scala.collection.immutable.Set[Int], Int, scala.collection.immutable.BitSet](scala.collection.immutable.BitSet.canBuildFrom)
breaker: scala.collection.generic.CanBuildFrom[scala.collection.immutable.Set[Int],Int,scala.collection.immutable.BitSet] = scala.collection.generic.BitSetFactory$$anon$1@3dcb28de

scala> val f = java.lang.Math.abs _
f: Int => Int = $$Lambda$1331/1191964420@1ab08e83

scala> Set[Int]().map[Int, scala.collection.immutable.BitSet](f)(breaker)
java.lang.ClassCastException: scala.collection.immutable.Set$EmptySet$ cannot be cast to scala.collection.BitSet
  at scala.collection.generic.BitSetFactory$$anon$1.apply(BitSetFactory.scala:36)
  at scala.collection.TraversableLike.builder$1(TraversableLike.scala:233)
  at scala.collection.TraversableLike.map(TraversableLike.scala:237)
  at scala.collection.TraversableLike.map$(TraversableLike.scala:231)
  at scala.collection.AbstractSet.scala$collection$SetLike$$super$map(Set.scala:51)
  at scala.collection.SetLike.map(SetLike.scala:104)
  at scala.collection.SetLike.map$(SetLike.scala:104)
  at scala.collection.AbstractSet.map(Set.scala:51)
```

**Comment by mkeskells** [comment]
rebased to resolve expected conflict with https://github.com/scala/scala/pull/8467

File: src/library/scala/collection/immutable/HashMap.scala
Comment (szeiger) [inline src/library/scala/collection/immutable/HashMap.scala]: I'm not sure if you can ever get multiple levels of wrapping but `isCompatibleCBF(w.unwrap)` would also be more concise and avoid repeating the test

File: src/library/scala/collection/immutable/HashMap.scala
Comment (szeiger) [inline src/library/scala/collection/immutable/HashMap.scala]: Is the separate method necessary? Couldn't you put the implementation directly into the second `++` method?

File: src/library/scala/collection/immutable/HashMap.scala
Comment (szeiger) [inline src/library/scala/collection/immutable/HashMap.scala:184]: Is the handling of key and value identities here and in `++:` below compatible with the current default (and new fallback) implementation consisting of successive `+` calls?

File: src/library/scala/collection/immutable/HashMap.scala
Comment (szeiger) [inline src/library/scala/collection/immutable/HashMap.scala]: Is the default merger (aka `liftMerger(null)`) different from a `null` merger (as used when calling `+`)?

File: src/library/scala/collection/immutable/HashMap.scala
Comment (szeiger) [inline src/library/scala/collection/immutable/HashMap.scala]: This whole branch appears to be identical to the one from the other `++:` method above. It could be factored out and shared.

File: src/library/scala/collection/immutable/HashMap.scala
Comment (szeiger) [inline src/library/scala/collection/immutable/HashMap.scala:281]: Probably, but the override wouldn't be binary compatible and there's a risk of breaking existing implementations that make `++` in `Map` call `MapBuilder`.

File: src/library/scala/collection/immutable/HashMap.scala
Comment (szeiger) [inline src/library/scala/collection/immutable/HashMap.scala]: Off-topic: I wonder if `unsignedCompare` (for `Int`) was ever benchmarked against a more straight-forward version that performs unsigned conversions to `Long`. It's also confusingly named (compare? for what?)

File: src/library/scala/collection/immutable/Map.scala
Comment (szeiger) [inline src/library/scala/collection/immutable/Map.scala]: It looks like this is only called from the first `++` above where the cbf instance is statically known.

File: src/library/scala/collection/package.scala
Comment (szeiger) [inline src/library/scala/collection/package.scala]: Why do we need this method that drops the `From` type instead of making `wrapped` a `val` and keeping `Nothing`?

File: src/library/scala/collection/immutable/HashMap.scala
Comment (mkeskells) [reply src/library/scala/collection/immutable/HashMap.scala]: done

File: src/library/scala/collection/immutable/HashMap.scala
Comment (mkeskells) [reply src/library/scala/collection/immutable/HashMap.scala]: if I chnage  to 
```
  override def ++[B1 >: B](xs: GenTraversableOnce[(A, B1)]): Map[A, B1] = ++(xs)(HashMap.canBuildFrom[A, B1])
```
I get a compile error
```
Error:(166, 102) type mismatch;
 found   : scala.collection.generic.CanBuildFrom[scala.collection.immutable.HashMap.Coll,(A, B1),scala.collection.immutable.HashMap[A,B1]]
    (which expands to)  scala.collection.generic.CanBuildFrom[scala.collection.immutable.HashMap[_, _],(A, B1),scala.collection.immutable.HashMap[A,B1]]
 required: A
  override def ++[B1 >: B](xs: GenTraversableOnce[(A, B1)]): Map[A, B1] = ++(xs)(HashMap.canBuildFrom[A, B1])
```

File: src/library/scala/collection/immutable/HashMap.scala
Comment (mkeskells) [reply src/library/scala/collection/immutable/HashMap.scala:281]: Maybe we could do something in 2.15, when that branch starts

File: src/library/scala/collection/immutable/HashMap.scala
Comment (mkeskells) [reply src/library/scala/collection/immutable/HashMap.scala]: I think (based in the HashSet changes, this can probably go from this implementation later

File: src/library/scala/collection/immutable/Map.scala
Comment (mkeskells) [reply src/library/scala/collection/immutable/Map.scala]: doh - changed to be similar to HashMap `++` and `addImpl`

File: src/library/scala/collection/package.scala
Comment (mkeskells) [reply src/library/scala/collection/package.scala]: no idea. Changed it to be a val

**Comment by mkeskells** [comment]
addressed some of @szeiger comments
rebased - (removed the mima changes - easier to start again on those from the build)

File: src/library/scala/collection/immutable/HashMap.scala
Comment (mkeskells) [reply src/library/scala/collection/immutable/HashMap.scala]: done - and removed the ObjectRef
TODO - should cope with the Maps that now have `forEachEntry`

File: src/library/scala/collection/immutable/HashMap.scala
Comment (mkeskells) [reply src/library/scala/collection/immutable/HashMap.scala]: yes - `liftMerger(null)` uses a merger with `retainIdentical = true` 

```
  private def liftMerger[A1, B1](mergef: MergeFunction[A1, B1]): Merger[A1, B1] =
    if (mergef == null) defaultMerger.asInstanceOf[Merger[A1, B1]] else liftMerger0(mergef)

  private val defaultMerger : Merger[Any, Any] = new Merger[Any, Any] {
    override def apply(a: (Any, Any), b: (Any, Any)): (Any, Any) = a
    override def retainIdentical: Boolean = true
    override val invert: Merger[Any, Any] = new Merger[Any, Any] {
      override def apply(a: (Any, Any), b: (Any, Any)): (Any, Any) = b
      override def retainIdentical: Boolean = true
      override def invert = defaultMerger
    }
  }
```

this means that the merge code can ignore structurally shared trees

other than that the behaviour is , or should be, identical

File: src/library/scala/collection/immutable/HashMap.scala
Comment (mkeskells) [reply src/library/scala/collection/immutable/HashMap.scala:184]: I believe so - that was my intention

**Comment by mkeskells** [comment]
rebased and squashed

File: src/library/scala/collection/immutable/HashMap.scala
Comment (szeiger) [reply src/library/scala/collection/immutable/HashMap.scala]: It works if you specify the type parameters:
```
  override def ++[B1 >: B](xs: GenTraversableOnce[(A, B1)]): Map[A, B1] = ++[(A, B1), Map[A, B1]](xs)(HashMap.canBuildFrom[A, B1])

  override def ++[C >: (A, B), That](that: GenTraversableOnce[C])(implicit bf: CanBuildFrom[HashMap[A, B], C, That]): That = {
```
The difference must be due to `++` being overloaded. It's strange that it fails with this type error rather than an ambiguous overload error.

Comment (mkeskells) [reply src/library/scala/collection/immutable/HashMap.scala]: done - thanks for the type magic :-)

File: src/library/scala/collection/immutable/HashMap.scala
Comment (mkeskells) [reply src/library/scala/collection/immutable/HashMap.scala]: did the forEachEntry change, and copied most of that to another PR to reduce the noise in the one

File: src/library/scala/collection/immutable/HashMap.scala
Comment (mkeskells) [reply src/library/scala/collection/immutable/HashMap.scala]: removed reference to unsignedCompare from HashMap. There are still some from HashSet until https://github.com/scala/scala/pull/8589 merges

File: src/library/scala/collection/immutable/HashMap.scala
Comment (szeiger) [inline src/library/scala/collection/immutable/HashMap.scala:219]: Is this really better than using `foreach`? Now you don't get a tuple anymore (even when the original map has one) so you have to compute a new one later on demand.

File: src/library/scala/collection/immutable/HashMap.scala
Comment (szeiger) [inline src/library/scala/collection/immutable/HashMap.scala]: Doesn't this need to be guarded by `merger.retainIdentical` but can be removed because it was already covered by the previous branch anyway?

File: src/library/scala/collection/immutable/HashMap.scala
Comment (mkeskells) [reply src/library/scala/collection/immutable/HashMap.scala:219]: Its a trade off. maybe the foreachEntry should have a tuple if it has one?
It depends if you generate the tuple, and how many Maps actually have the tuple I think

File: src/library/scala/collection/immutable/HashMap.scala
Comment (mkeskells) [reply src/library/scala/collection/immutable/HashMap.scala]: I don't see why. 
If the merge generates a tuple which is identical in all respects then we don't need to create a new `HashMap1`, when this is a clone and identical in all respects to `this`
What user visible difference is there? 
Non-functionally there is a change because not having this would cause the creation of a new `HashMap1` and all of the parent structures, and the loss of structural sharing

File: src/library/scala/collection/immutable/HashMap.scala
Comment (mkeskells) [reply src/library/scala/collection/immutable/HashMap.scala:219]: the only classes that implement `HasForEachEntry` are 
`HashMap` - covered previously in matches
`ListMap` - doesn't  maintain tuples
`SortedMap`/`RedBlackTree` - doesn't maintain tuples 

which collections types are you concerned with, when we have the existing tuple and lose it?

**Comment by retronym** [comment]
Pushed a commit to rebase atop the latest 2.12.x (which has `HasForEach`) and to squash the last two commits.

**Comment by retronym** [comment]
/nothingtoseehere

(Clearing spurious "cla Expected — Waiting for status to be reported ")

**Review by szeiger — APPROVED** [review]

File: src/library/scala/collection/immutable/HashMap.scala
Comment (mkeskells) [reply src/library/scala/collection/immutable/HashMap.scala]: I think I see the confusion - there was a bug here
key and values are updated to be this.key and this.value - no referring to the parameters

Comment (szeiger) [reply src/library/scala/collection/immutable/HashMap.scala]: Ah, that explains it. I was looking for the context of the code again after your previous comment but github didn't have it anymore so I looked at the latest version and everything there looked fine.

**Comment by szeiger** [comment]
LGTM but it's hard to judge the correctness of all the new code so I'd prefer to get a 2nd approval from @retronym before merging.

**Review by retronym — APPROVED** [review]

**Comment by SethTisue** [comment]
after this PR, scala-ide/scalariform has test failures in the community build: https://github.com/scala/scala-dev/issues/672 . let's discuss further there

**Comment by retronym** [comment]
Fixed in https://github.com/scala/scala/pull/8764

**Pull request merged**
